keps/sig-cli/kustomize-subcommand-integration.md

```diff
@@ -0,0 +1,387 @@
+invoke kubectl's patch and apply logic from a script (add-on manager circa 2016).
+
+Users have asked for a declarative script-free way to apply changes to Resource Config.
+
+### Goals
+
+Solve common kubectl user friction (such as those defined in the Motivation section) by publishing kustomize
+functionality from kubectl to complement commands which are targeted at declarative workflows.
+
+- Complement commands targeted at declarative workflows: `kubectl apply`, `kubectl diff`
+- Complement the *server-side apply* initiative targeted at improving declarative workflows.
+
+User friction solved through capabilities such as:
+
+- Generating Resource Config for Resources with data frequently populated from other sources - ConfigMap and Secret
+- Performing common cross-cutting transformations intended to be applied across Resource Configs - e.g.
+  name prefixing / suffixing, labeling, annotating, namespacing, imageTag setting.
+- Performing flexible targeted transformations intended to be applied to specific Resource Configs - e.g.
+  strategic patch, json patch
+- Composing and layering Resource Config with schema-aware transformations
+- Facilitating rolling updates to Resources such as ConfigMaps and Secrets via Resource Config transformations
+- Facilitating creation of Resources that require the creation to be ordered - e.g. namespaces, crds
+
+### Non-Goals
+
+- Exposing all imperative kubectl commands as declarative options
+- Exposing all kubectl generators and schema-aware transformations
+- Providing simpler alternatives to the APIs for declaring Resource Config (e.g. a simple way to create deployments)
+- Providing a templating or general substitution mechanism (e.g. for generating Resource Config)
+
+### Why should this be part of kubectl?
+
+- It was purposefully built to address user friction in kubectl declarative workflows. Leaving these issues unaddressed
+  in the command itself reduces the quality of the product.
+- The techniques it uses to address the issues are based on the existing kubectl imperative commands.  It is
+  consistent with the rest of kubectl.
+- Bridging the imperative and declarative workflows helps bring a more integrated feel and consistent story to
+  kubectl's command set.
+- Kustomize is already part of the Kubernetes project and owned by SIG CLI (the same sig that owns kubectl).
+  SIG CLI members have expertise in the kustomize codebase and are committed to maintaining the solution going forward.
+- Providing it as part of kubectl will ensure that it is available to users of kubectl apply and simplify the
+  getting started experience.
+
+## Proposal
+
+Kustomize Standalone Sub Command
+
+Publish the `kustomize build` command as `kubectl kustomize`.  Update documentation demonstrate using kustomize as
+`kubectl kustomize <dir> | kubectl apply -f -`.
+
+`kubectl kustomize` takes a single argument with is the location of a directory containing a file named `Kustomization`
```

Comment:
file named `kustomization.yaml`

Reply:
@monopole WDYT of changing the file name pre-integration (we could still recognize the old one, but not advertise it)?  This follows the pattern of `Makefile` and `Dockerfile`.

Reply:
FWIW, We went with `.yaml` (and, alas, `.yml`) because emacs and vi both still use the legacy convention as a pretty reliable edit model signal.  I just tried it without the extension, and my stock emacs and vim installs failed to do the Right Thing based on content detection.

`Makefile`, being unique text formatting classes with one instantiation each, doesn't suffer from this problem (emacs/vi do the Right Thing).  `Dockerfile` detection hasn't made it into my stock emacs install yet :)

I'm for allowing all of them (pre integration), in the precedence `Kustomization`, `kustomization.yaml`, `kustomization.yml`, erroring only if none can be found or if more than one found.

Reply:
Yes we have.

Reply:
keeping this `kustomization.yaml`

Reply:
Honestly, since we're introducing `-k` flag which is meant to point to `kustomization.yaml` I think we can safely assume that any name is supported, right? Or did I miss something? Eventually we should make this more clearer in description of the flag, if so. But I remember from the discussion we've had during KubeCon we wanted to explicitly point to the file to run kustomization, unless the fact that we're using `-k/--kustomize` allows users to pass file OR the directory containing the file. In which case there are two possibilities:
1. if I pass file name to `-k` the name doesn't matter, only the kind specified in file must be `Kustomization`.
2. if I pass directory we can search for `kustomization.yaml` or just look for a file with `Kustomization` kind.
So actually the name doesn't matter. Other thoughts?

Reply:
I'd expect -k to take a directory, and look for the kustomization file the same way we look for the file when processing a base referenced from another file. For these directories to be composable, they will need to use kustomize filenames findable via base reference, anyway.

Reply:
That's a reasonable argument for `-k` accepting a dir.